CONTRIBUTING.md

```diff
@@ -0,0 +1,81 @@
+
+Any help pushing forward the `scie-pants` launcher and its supporting Python tools is very welcome.
+Thank you in advance for your time and effort.
+
+## Development Environment
+
+You'll need just a few tools to hack on scie-pants:
++ If you're new to Rust you'll need to ensure you have its toolchain available to you. You might
+  start with [`rustup`](https://rustup.rs/).
++ Currently, bootstrapping builds [ptex](https://github.com/a-scie/ptex) to use in fetching
+  additional build dependencies. Although `ptex` is a Rust project as well, it requires [CMake](
+  https://cmake.org/) for one of its sys crates. CMake is generally available via your operating
+  system's package manager or Homebrew, for example, if you're on Mac.
++ The [`scie-jump`](https://github.com/a-scie/jump) launcher plays a critical role in the structure
+  and operation of `scie-pants`. Although the [package](package/src/main.rs) build process takes
+  care of obtaining the `scie-jump` binary itself, you should familiarize yourself with its
+  [packaging format](https://github.com/a-scie/jump/blob/main/docs/packaging.md), which both the
+  [build process itself](package/pbt.lift.json) and the final [`scie-pants` binary](
+  package/scie-pants.lift.json) use.
+
+## Development Cycle
+
+You might want to open a [discussion](https://github.com/pantsbuild/scie-pants/discussions) or
+[issue](https://github.com/pantsbuild/scie-pants/issues) to vet your idea first. It can often save
+overall effort and lead to a better end result.
+
+The code is run through the ~standard `cargo` gamut. Before sending off changes you should have:
++ Formatted the code (requires Rust nightly): `cargo +nightly fmt --all`
++ Linted the code: `cargo clippy --all`
++ Tested the code: `cargo test --all`
+
+Additionally, you can run any existing integration tests with `cargo run -p package -- test`. This
+packages the `scie-pants` scie and then uses it to launch Pants which formats, lints, checks, tests
+and re-packages the scie-pants [tools](tools) Python support code.
+
+You can also just package the `scie-pants` scie binary via `cargo run -p package -- scie`. That will
+build the `scie-pants` binary for the current machine to the `dist/` directory by default (run
+`cargo run -p package -- --help` to find out more options). Two files will be produced there:
+1. The scie jump binary: `scie-pants-<os>-<arch>(.<ext>)`
+2. The scie jump fingerprint file: `scie-pants-<os>-<arch>(.<ext>).sha256`
+
+You can then run `dist/scie-pants-<os>-<arch>(.<ext>) <pants goals>` to run Pants against the tools
+code when iterating on it.
+
+When you're ready to get additional eyes on your changes, submit a [pull request](
+https://github.com/pantsbuild/scie-pants/pulls).
+
+## Guiding Principles
+
+There are just a few guiding principles to keep in mind as alluded to above:
++ The `scie-pants` hot path should be fast: It currently launches Pants in ~one millisecond and that
+  should hold steady.
```

Comment:
Could we have/are there any perf tests that may run automatically/in CI to ensure this will hold true over time?

Reply:
We should. I'm not sure exactly how to do it yet. Pants gets in the way since its slow. There are 3 separate binary executions in the hot path:
1. The scie-jump launches the embedded scie-pants.bin binary as the default command.
2. The scie-pants binary finds pants.toml, parses the pants_version and determines the proper Python version and then re-launches itself specifying a few env vars to convey what it found and SCIE_BOOT=pants to launch the pants command contained in the scie.
3. The scie-jump launches the pants command.
4. The pants console script in the previously installed Pants venv runs.

Each of the 1st 3 steps takes ~300us. The 4th step takes 500ms min (for pants -V for example).

Perhaps there is a way to shim out 4 with something other than the pants console script that runs in consistent time on the order of 1ms or less (like `bash -V`). Perhaps even that could be avoided.

Right now I just run with RUST_LOG=debug which gives the timings I just described for the 1st 3 steps in log lines.

Reply:
#24 now tracks this.

Reply:
I will say though to not really worry about this if it feels scary when considering contribution. This is both because the actual scie-pants binary does almost nothing; so its hard to make it slow, and because most feature work will just be, at most, plumbing reading of new toml values from an already priced in pants.toml parse and setting new env vars for sub-processes - both of which are really hard to make slow in non-obvious ways.

Reply:
The slowest thing it does actually, by almost an order of magnitude, is handle `.pants.bootstrap`! So this is step 2 which takes ~300us with no `.pants.bootstrap` and adding in a ~minimal GIT_COMMIT `./pants.bootstrap` adds >2000us.